ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/FuncLongToDecimal64.java

```diff
@@ -0,0 +1,185 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.exec.vector.expressions;
+
+import java.util.Arrays;
+
+import org.apache.hadoop.hive.ql.exec.vector.Decimal64ColumnVector;
+import org.apache.hadoop.hive.ql.exec.vector.LongColumnVector;
+import org.apache.hadoop.hive.ql.exec.vector.VectorExpressionDescriptor;
+import org.apache.hadoop.hive.ql.exec.vector.VectorizedExpressionsSupportDecimal64;
+import org.apache.hadoop.hive.ql.exec.vector.VectorizedRowBatch;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.serde2.typeinfo.DecimalTypeInfo;
+
+/**
+ * This is a superclass for unary long functions and expressions returning decimals that
+ * operate directly on the input and set the output.
+ */
+@VectorizedExpressionsSupportDecimal64()
+public abstract class FuncLongToDecimal64 extends VectorExpression {
+  private static final long serialVersionUID = 1L;
+  private final int inputColumn;
+
+  private static final long[] powerOfTenTable = {
+    1L,                   // 0
+    10L,
+    100L,
+    1_000L,
+    10_000L,
+    100_000L,
+    1_000_000L,
+    10_000_000L,
+    100_000_000L,           // 8
+    1_000_000_000L,
+    10_000_000_000L,
+    100_000_000_000L,
+    1_000_000_000_000L,
+    10_000_000_000_000L,
+    100_000_000_000_000L,
+    1_000_000_000_000_000L,
+    10_000_000_000_000_000L,   // 16
+    100_000_000_000_000_000L,
+    1_000_000_000_000_000_000L, // 18
+  };
+
+  public FuncLongToDecimal64(int inputColumn, int outputColumnNum) {
+    super(outputColumnNum);
+    this.inputColumn = inputColumn;
+  }
+
+  public FuncLongToDecimal64() {
+    super();
+
+    // Dummy final assignments.
+    inputColumn = -1;
+  }
+
+  abstract protected void func(Decimal64ColumnVector outputColVector, LongColumnVector inputColVector, int i,
+      long scaleFactor);
+
+  @Override
+  public void evaluate(VectorizedRowBatch batch) throws HiveException {
+
+    if (childExpressions != null) {
+      super.evaluateChildren(batch);
+    }
+
+    LongColumnVector inputColVector = (LongColumnVector) batch.cols[inputColumn];
+    int[] sel = batch.selected;
+    int n = batch.size;
+    Decimal64ColumnVector outputColVector = (Decimal64ColumnVector) batch.cols[outputColumnNum];
+    int outputScale = ((DecimalTypeInfo) outputTypeInfo).scale();
+    final long scaleFactor = powerOfTenTable[outputScale];
```

Comment:
Is outputScale always guaranteed to be between [0, 18]?

Reply:
Yes, before we create cast expression, we verify if the output.precision + output.scale <= 18(to avoid intermediate overflow) and hence implicitly outputscale will be less than or equal to 18.